### ISSUE TITLE
[RFC] Taichi on Javascript

### ISSUE THREAD

**Issue by AmesingFlank** [taichi-dev/taichi#3781]
I'm writing this RFC to share my thoughts and plans to build a version of Taichi completely embedded in Javascript, with the hope of getting production-level applications to use Taichi, and thereby increasing the impact of Taichi as a programming language.

Currently, Taichi uses Python as the host language. Thanks to Python's simplicity, Taichi enables programmers with little GPU experiences to have fun and experiment with GPU computing. Javascript shares the simplicity of Python (to some extent), and additionally, it has the advantage of being used extensively in production-level, user-facing applications. For this reason, if we offer Taichi as a (deployable) Javascript package, it would provide professional web developers with a brand new way of building and serving compute-heavy or graphics-heavy web programs. This would certainly boost the popularity of Taichi.

To be more specifically, I'm planning to build a Javascript library that enables this:

```javascript
const ti = require("taichi")

x = ti.Vector.field(3, ti.f32, 512)

let k = ti.kernel(() => {
    for (i of ti.range(512)) {
        x[i] = x[i] + 1
    }
})

k()
```

Here, `taichi` would be a normal Javascript module, one that could be installed simply by running  `npm install taichi`. Most importantly, it must be a pure Javascript module, meaning that it must have **no binary code dependencies** attached to it. This is of course vital if we want web developers to be able to use it in their own products. We will of course implement interop routines with normal Javascript arrays, HTML canvases, and maybe even provide a GGUI. 

# Use cases and Impacts

From the top of my head I can think of a lot of use cool cases for a Javascript version of Taichi

* Graphics applications on the web. 
No doubt there are lots and lots of graphics demands on web pages, and these are done most often via WebGL, which is flexible but cumbersome. As a language directly embedded in JS, this could be very appealing for beginner-level front-end engineers who don’t want to learn WebGL.

* Compute heavy applications on the web
These are considerably rarer, but part of the reason for that is exactly because it’s difficult to develop high performance programs on the web -- and taichi just might change that.

* Other fancy apps:
    * A (much better) taichi zoo, where actually everything happens on the frontend, and the backend server does nothing but serving (instead of compiling).
    * A compute resource sharing platform:
This is something I thought about a few years ago. We could build a site where people can submit compute tasks written in taichi, and other people could simple open up the website and contribute their (GPU) computing resources.
    * A transpiler between python-taichi and JS-taichi, which helps improve the impact of the python language.
 

# Technological Considerations and Challenges

Obviously, this will be a huge project with many challenges. Here’re the things that have crossed my mind:

* Feasibility on the Javascript side
In Javascript, you can [obtain the source code](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/toString) of a function passed in as an argument. So this shouldn’t be a problem. In fact, there is a library called gpu.js[https://gpu.rocks/#/] which already uses this to translate Javascript to GPU code. That package is pretty cool, but Taichi is much, much more powerful of course.
Also, Javascript might [support decorators](https://medium.com/jspoint/a-minimal-guide-to-ecmascript-decorators-55b70338215e) in the future. If this ECMAScript proposal goes through, we might be able to have
```javascript
const ti = require("taichi")

x = ti.Vector.field(3, ti.f32, 512)

@ti.kernel
function k() {
    for (i of ti.range(512)) {
        x[i] = x[i] + 1
    }
}

k()
```

* No binary code dependencies?
Emscripten. I have already successfully compiled the entire Taichi project using emscripten. I haven't started trying to use the generated code, and there will probably be some troubles to overcome, but overall this is very feasible.

* GPU Computing
This is a big "X" of this project. WebGL doesn’t have good support for compute. WebGPU does, but browsers don’t have good support for WebGPU. If we only use CPU, e.g. WASM, then the performance probably won’t be enough of a selling point.
Although, I remain hopeful that not too long in the future, WebGPU will become widely adopted. I mean, WebGL can’t just live forever right? 
<p align="center">
  <img  width="300" src="https://user-images.githubusercontent.com/16066115/145727484-9c6693f7-a271-48da-bfaf-13786142557c.png" />
</p>

# Roadmap
* [ ] Javascript Frontend. ( i.e. JS source code -> CHI IR)
I believe the best way to do this is to use the [Typescript Compiler API](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API).
* [ ] Fully emscripten Taichi
After the Javascript frontend, we will use a emscriptened Taichi to translate CHI IR to SPIR-V
* [ ] SPIR-V to WGSL.
There are a few publicly available tools that does this, namely [naga](https://github.com/gfx-rs/naga) and [tint](https://dawn.googlesource.com/tint/). But in fact none of them works well enough and there're some painful problems. I won't go into the details here, instead, I will post a comment below to discuss this.
* [ ] A WebGPU runtime that injests the WGSL.

# Proof of Concept

I have hacked up a WebGPU Taichi runtime and used Taichi AOT and Tint to generate some WGSL kernels. If you have Chrome 94+, you can see it here:

https://amesingflank.github.io/ti.js/

**Comment by AmesingFlank** [comment]
Regarding SPIR-V to WGSL, the biggest pain point is atomics. To understand the problem, please have a look at [this discussion](https://github.com/gpuweb/gpuweb/issues/2377).

TLDR:
* In WGSL, atomic operations can only be applied to `atomic<T>` types. And there is no way to convert a `T*` to a `atomic<T>*`. 
* If we wish to perform any atomic operations at all on a buffer, we would need to promote the entire buffer into atomic types, and always use atomic operations (even load/store). This not only inconvenient, but also inefficient.
* Neither tint or naga handles this. They immediately give up when seeing an atomic.

I'd appreciate any help/suggestions on this, as this feels like a huge blocker. To make matters worse, it looks that the WebGPU committee has no plans to make atomics more flexible in the near future, as they have moved the discussion to `post-v1`.

**Comment by bobcao3** [comment]
> Regarding SPIR-V to WGSL, the biggest pain point is atomics. To understand the problem, please have a look at [this discussion](https://github.com/gpuweb/gpuweb/issues/2377).
> 
> TLDR:
> 
>     * In WGSL, atomic operations can only be applied to `atomic<T>` types. And there is no way to convert a `T*` to a `atomic<T>*`.
> 
>     * If we wish to perform any atomic operations at all on a buffer, we would need to promote the entire buffer into atomic types, and always use atomic operations (even load/store). This not only inconvenient, but also inefficient.
> 
>     * Neither tint or naga handles this. They immediately give up when seeing an atomic.
> 
> 
> I'd appreciate any help/suggestions on this, as this feels like a huge blocker. To make matters worse, it looks that the WebGPU committee has no plans to make atomics more flexible in the near future, as they have moved the discussion to `post-v1`.

WebGPU is not a stable API by any means and the atomic issue also originates from their persistence to use WGSL instead of a common standard. This does not sound like an addressable issue in any way, we might want to straight up limit atomic capabilities on WGSL all together. This is my personal opinion, but I feel like if the platform did not allow flexibility, we should not try to "fix" it by hacking our ways around it

**Comment by bobcao3** [comment]
Would it be reasonable to make this work two-parted. The first part is the front-end, and we can use bindings such as https://github.com/maierfelix/nvk to allow at least node.js front-end to work with Vulkan. The second part would be the backend where we figure out WebGPU and WGSL

**Comment by AmesingFlank** [comment]
> This does not sound like an addressable issue in any way, we might want to straight up limit atomic capabilities on WGSL all together. This is my personal opinion, but I feel like if the platform did not allow flexibility, we should not try to "fix" it by hacking our ways around it

I more or less agree. I hate the idea of hacking around it either. But this would essentially mean we would support **no** atomics at all, and that's not great. Virtually all of the physical simulation programs in taichi requires atomics. Not having atomics is a huge limitation. 

> Would it be reasonable to make this work two-parted. The first part is the front-end, and we can use bindings such as https://github.com/maierfelix/nvk to allow at least node.js front-end to work with Vulkan. The second part would be the backend where we figure out WebGPU and WGSL

That makes sense. Although I wouldn't invest too much time into a node-only version if it turns out to be a lot of work. My reasoning being that a node-only version doesn't add that much value to our existing Python version, while a version that runs in browsers can open up a lot of possibilities.

**Comment by eatcosmos** [comment]
https://amesingflank.github.io/taichi.js/
compile and run not work?

**Comment by AmesingFlank** [comment]
> https://amesingflank.github.io/taichi.js/ compile and run not work?

Hi, thanks for reporting. Could you please tell me your browser version? And could you paste the code that you put in the code editor? Are there any error messages in the Logs?

**Comment by eatcosmos** [comment]
> > https://amesingflank.github.io/taichi.js/ compile and run not work?
> 
> Hi, thanks for reporting. Could you please tell me your browser version? And could you paste the code that you put in the code editor? Are there any error messages in the Logs?

96.0.4664.45（） （32 ）
![image](https://user-images.githubusercontent.com/83876321/155984669-7cd4e2bc-0a1c-4d29-84f4-0db6e46fdd66.png)
```
19 error(s) generated while compiling the shader:
27:28 error: float literals must not be suffixed with 'f'
  let tmp1 : f32 = 1.000000f;
                           ^

31:28 error: float literals must not be suffixed with 'f'
  let tmp5 : f32 = 0.150000f;
                           ^

32:28 error: float literals must not be suffixed with 'f'
  let tmp6 : f32 = 0.450000f;
                           ^

33:28 error: float literals must not be suffixed with 'f'
  let tmp7 : f32 = 0.500000f;
                           ^

34:28 error: float literals must not be suffixed with 'f'
  let tmp8 : f32 = 3.900000f;
                           ^

35:28 error: float literals must not be suffixed with 'f'
  let tmp9 : f32 = 1.400000f;
                           ^

36:29 error: float literals must not be suffixed with 'f'
  let tmp10 : f32 = 1.100000f;
                            ^

37:29 error: float literals must not be suffixed with 'f'
  let tmp11 : f32 = 1.200000f;
                            ^

38:29 error: float literals must not be suffixed with 'f'
  let tmp12 : f32 = 1.700000f;
                            ^

39:29 error: float literals must not be suffixed with 'f'
  let tmp13 : f32 = 1.300000f;
                            ^

40:29 error: float literals must not be suffixed with 'f'
  let tmp14 : f32 = 2.500000f;
                            ^

67:29 error: float literals must not be suffixed with 'f'
  let tmp41 : f32 = 0.300000f;
                            ^

71:29 error: float literals must not be suffixed with 'f'
  let tmp45 : f32 = 1.800000f;
                            ^

72:29 error: float literals must not be suffixed with 'f'
  let tmp46 : f32 = 0.370000f;
                            ^

78:29 error: float literals must not be suffixed with 'f'
  let tmp52 : f32 = 0.800000f;
                            ^

80:29 error: float literals must not be suffixed with 'f'
  let tmp54 : f32 = 0.330000f;
                            ^

87:29 error: float literals must not be suffixed with 'f'
  let tmp61 : f32 = 2.200000f;
                            ^

88:29 error: float literals must not be suffixed with 'f'
  let tmp62 : f32 = 0.310000f;
                            ^

112:29 error: float literals must not be suffixed with 'f'
  let tmp86 : f32 = 0.100000f;
                            ^

index.html#:1 6 error(s) generated while compiling the shader:
39:31 error: float literals must not be suffixed with 'f'
    let tmp99 : f32 = 0.000000f;
                              ^

40:32 error: float literals must not be suffixed with 'f'
    let tmp100 : f32 = 0.769231f;
                               ^

42:32 error: float literals must not be suffixed with 'f'
    let tmp102 : f32 = 1.000000f;
                               ^

43:32 error: float literals must not be suffixed with 'f'
    let tmp103 : f32 = 0.000100f;
                               ^

44:32 error: float literals must not be suffixed with 'f'
    let tmp104 : f32 = 0.250000f;
                               ^

70:32 error: float literals must not be suffixed with 'f'
    let tmp123 : f32 = 0.750000f;
                               ^

```

**Comment by AmesingFlank** [comment]
@eatcosmos Thanks! I will try to fix this.

**Comment by AmesingFlank** [comment]
@eatcosmos Also, could you try upgrading to chrome v98?

**Comment by eatcosmos** [comment]
> @eatcosmos Also, could you try upgrading to chrome v98?

v98 is ok, load image quickly

**Comment by eatcosmos** [comment]
What about webgl Oasis Engines, is three any relation between Taichi.js and Oasis or Three.js in technology？
[Oasis Engine \- Mobile first web graphic engine](https://oasisengine.cn/)
[支付宝App 3D动画和小游戏背后的故事\_前端\_曾柏然\_InfoQ精选文章](https://www.infoq.cn/article/w1f7zziah1aud5dp6kui)

**Comment by bobcao3** [comment]
> What about webgl Oasis Engines, is three any relation between Taichi.js and Oasis or Three.js in technology？ [Oasis Engine - Mobile first web graphic engine](https://oasisengine.cn/) [支付宝App 3D动画和小游戏背后的故事_前端_曾柏然_InfoQ精选文章](https://www.infoq.cn/article/w1f7zziah1aud5dp6kui)

There isn't much in common at all. Taichi.js is a compiler codegen and JIT runtime system that compiles your JS code into GPU kernels. These examples are graphics and rendering engines, they are very different things. So, no relation in them at all.

**Comment by eatcosmos** [comment]
> > What about webgl Oasis Engines, is three any relation between Taichi.js and Oasis or Three.js in technology？ [Oasis Engine - Mobile first web graphic engine](https://oasisengine.cn/) [支付宝App 3D动画和小游戏背后的故事_前端_曾柏然_InfoQ精选文章](https://www.infoq.cn/article/w1f7zziah1aud5dp6kui)

>

> There isn't much in common at all. Taichi.js is a compiler codegen and JIT runtime system that compiles your JS code into GPU kernels. These examples are graphics and rendering engines, they are very different things. So, no relation in them at all.

| run [taichi.js demo](https://amesingflank.github.io/taichi.js/zoo/index.html) in mobile chrome | run [three.js demo](https://threejs.org/examples/#webgl_animation_skinning_morph) in mobile chrome | run [oasis.js demo](https://oasisengine.cn/0.6/examples) in mobile chrome | 
|---|---|---|
| ![2](https://raw.githubusercontent.com/eatcosmos/assets/master/2.7bfdoa6b36g0.jpg)  | ![1](https://raw.githubusercontent.com/eatcosmos/assets/master/1.jpg)  | ![3](https://raw.githubusercontent.com/eatcosmos/assets/master/3.3x3ks4mvtye0.jpg)  |

Test taichi.js three.js oasis.js, taichi.js cannot show demo in mobile chrome, 
what need to do to let taichi.js run directly in mobile chrome? 
Is this feature is taichi.js's objective in the future?

**Comment by AmesingFlank** [comment]
@eatcosmos Taichi.js runs on WebGPU, which is not supported by mobile chrome yet. 

Of course I hope that in the future, mobile chrome/safari will support WebGPU and thus Taichi.js. But at this point there's no telling how long that would take.

**Comment by eatcosmos** [comment]
> @eatcosmos Taichi.js runs on WebGPU, which is not supported by mobile chrome yet.
> 
> Of course I hope that in the future, mobile chrome/safari will support WebGPU and thus Taichi.js. But at this point there's no telling how long that would take.

but mobile has no gpu, three.js could work, three.js use webgl , taichi.js not use webgl? 
Then will taichi.js do the webgl's work , so taichi.js can work in mobile chrome without using webgl?

**Comment by AmesingFlank** [comment]
@eatcosmos Three.js uses WebGL, which is a legacy API, and is well-supported by mobile browsers. Taichi.js uses a new API called WebGPU, which is aimed to replace WebGL, but has a long way to go from actually doing so.

**Comment by k-ye** [comment]
FYI you can track Chrome's support status of WebGPU in https://chromestatus.com/feature/6213121689518080. It doesn't mention mobile Chrome yet, but this is almost a guaranteed thing.

**Comment by AmesingFlank** [comment]
> 

We use emscripten to compile taichi's C++ code base into javascript, paired with a WebGPU runtime written in Typescript.

**Comment by eatcosmos** [comment]
thank you for your explanation，maybe webgl is also a good choice for now to cover most devices

**Comment by AmesingFlank** [comment]
The problem with WebGL is that it doesn't support compute shaders. It is a legacy API and we expect that in due time,  WebGPU will receive fairly common support.

**Comment by sridhar-mani** [comment]
is there any attempt going on to implement taichi js in react native too
